bin/light-base/src/runtime_service.rs

```diff
@@ -765,30 +371,35 @@ impl<TPlat: Platform> RuntimeService<TPlat> {
         block_hash: &[u8; 32],
     ) {
         let mut guarded_lock = guarded.lock().await;
+        let guarded_lock = &mut *guarded_lock;
 
-        if guarded_lock
-            .pinned_blocks
-            .remove(&(subscription_id.0, *block_hash))
-            .is_none()
+        if let GuardedInner::FinalizedBlockRuntimeKnown {
+            all_blocks_subscriptions,
+            pinned_blocks,
+            ..
+        } = &mut guarded_lock.tree
         {
-            // Cold path.
-            if guarded_lock
-                .all_blocks_subscriptions
-                .contains_key(&subscription_id.0)
+            if pinned_blocks
+                .remove(&(subscription_id.0, *block_hash))
+                .is_none()
             {
-                panic!("block already unpinned");
-            } else {
-                return;
-            }
-        };
+                // Cold path.
+                if all_blocks_subscriptions.contains_key(&subscription_id.0) {
+                    panic!("block already unpinned");
```

Comment:
is there a guidance around the use of panics in the code? Can one use them even if they are not terminal (i.e. state isn't corrupted | program can continue to run)?

Reply:
The guidance is to use them when the state of the program would be corrupted. And that's the case here: while the state of the `runtime_service` itself isn't corrupted, the state of the stuff on top most likely is. Unpinning the same block twice is a bit like deleting the same object twice.